File: FileTreeWalker.scala
import java.io.File

object FileTreeWalker {
  def walkFileTree(f: File)(proc: File => Unit): Unit = {
    if (f.isDirectory) {
      f.listFiles foreach {child =>
        walkFileTree(child)(proc)
      }
    } else {
      proc(f)
    }
  }

  def main(args: Array[String]): Unit = {
    walkFileTree(new File("""C:\""")) {f =>
      println(f.getAbsolutePath)
    }
  }
}

File: FileTreeWalker.java
import java.io.File;
import java.util.Arrays;
import java.util.function.Consumer;

class FileTreeWalker {
  public static void walkFileTree(File f, Consumer<File> proc) {
    if (f.isDirectory()) {
      Arrays.stream(f.listFiles())
        .forEach(child -> walkFileTree(child, proc));
    } else {
      proc.accept(f);
    }
  }

  public static void main(String[] args) {
    walkFileTree(new File("C:\\"),
      f -> System.out.println(f.getAbsolutePath()));
  }
}
